src/d3/xy-chart.d3.ts

```diff
@@ -1,78 +1,76 @@
 import * as d3 from 'd3';
-import { BaseD3 } from './base.d3';
+import { BaseD3, RenderOptions as BaseRenderOptions } from './base.d3';
+import { Observable } from 'rxjs';
+import { formatX } from '../utils/formatters';
 
 export interface Datum {
   date: Date;
   value: number;
 }
 
-export interface RenderOptions {
-  data: Datum[];
-  height: number;
-  width: number;
-  marginTop: number;
-  marginRight: number;
-  marginBottom: number;
-  marginLeft: number;
+export interface RenderOptions extends BaseRenderOptions {
+  dataObservable: Observable<Datum[]>;
+  activeDatumObservable: Observable<Datum | null>;
 }
 
-export abstract class XYChartD3 extends BaseD3<RenderOptions> {
-  protected render({
-                     data,
-                     height, width,
-                     marginTop, marginRight, marginBottom, marginLeft,
-                   }: RenderOptions) {
-    const svg = this.container
-      .append('svg')
-      .attr('viewBox', [0, 0, width, height].join(' '));
-
-    const scaleX = d3
+export abstract class XYChartD3 extends BaseD3 {
+  protected scaleX: d3.ScaleTime<number, number>;
+  protected scaleY: d3.ScaleLinear<number, number>;
+
+  render(renderOptions: RenderOptions) {
+    super.render(renderOptions);
+
+    const {
+      dataObservable,
+      height, width,
+      marginTop, marginRight, marginBottom, marginLeft,
+    } = renderOptions;
+
+    this.scaleX = d3
       .scaleUtc()
-      .domain(d3.extent<Datum, Date>(data, d => d.date) as [Date, Date])
       .range([marginLeft, width - marginRight]);
 
-    const scaleY = d3
+    this.scaleY = d3
       .scaleLinear()
-      .domain([0, d3.max(data, d => d.value)!])
       .nice()
       .range([height - marginBottom, marginTop]);
 
-    svg
+    const xAxis = d3
+      .axisBottom<Date>(this.scaleX)
+      .ticks(width / 80)
+      .tickFormat(formatX)
+      .tickSizeOuter(0);
+
+    const yAxis = d3.axisLeft<number>(this.scaleY);
+
+    const xAxisG = this.svg
       .append('g')
-      .attr('transform', `translate(0,${height - marginBottom})`)
-      .call(
-        d3
-          .axisBottom(scaleX)
-          .ticks(width / 80)
-          .tickSizeOuter(0),
-      );
-
-    svg
+      .attr('transform', `translate(0,${height - marginBottom})`);
+
+    const yAxisG = this.svg
       .append('g')
-      .attr('transform', `translate(${marginLeft},0)`)
-      .call(d3.axisLeft(scaleY))
-      // .call(g => g.select('.domain').remove())
-      .call(g =>
-        g
-          .select('.tick:last-of-type text')
-          .clone()
-          .attr('x', 3)
-          .attr('text-anchor', 'start')
-          .attr('font-weight', 'bold')
-          .text('Active Users'),
-      );
-
-    this.appendChart(svg, data, scaleX, scaleY);
-
-    return () => {
-      svg.remove();
-    };
+      .attr('transform', `translate(${marginLeft},0)`);
+
+    dataObservable
```

Comment:
I'm probably being a bit picky here, but I think it might be easier to understand the render flow if this method contains the only subscribe to the dataObservable. Here's what I imagine it would look like:

```
render() {
  this.dataObservable
    .pipe(this.takeUntilCleared())
    .subscribe(data => {
      this.renderAxis(data);
      this.renderData(data);
      this.renderActiveDatum(data);
    });

  this.activeDataObservable
    .pipe(this.takeUntilCleared())
    .subscribe(datum => {
      this.renderActiveDatum(datum)
      // Could contain other actions based on activeDatum
    });
}
```

Tracking one subscription (or one for each observable) seems cleaner and potentially less error-prone to me. Also, would it be easier to store the RenderOptions as an instance variable on the base class, rather than passing them through these methods?

WDYT?